src/lua/init.c

```diff
@@ -433,19 +434,31 @@ static const char *lua_modules_preload[] = {
 static const char *
 tarantool_debug_getsources(const char *modname)
 {
-	for (size_t i = 0; lua_modules[i] != NULL; i += 2) {
-		const char *shortname = lua_modules[i];
-		const char *lua_code = lua_modules[i + 1];
-		assert(lua_code != NULL);
-		assert(strlen(shortname) + sizeof("@builtin/.lua") <=
-			MAX_MODNAME);
-		char fullname[MAX_MODNAME];
-		snprintf(fullname, sizeof(fullname), "@builtin/%s.lua",
-			 shortname);
-		if (!strcmp(shortname, modname) || !strcmp(fullname, modname))
-			return lua_code;
-	}
-	return NULL;
+	/* 1. process as short modname - fast path */
+	const char *lua_code = builtin_modcache_find(modname);
+	if (lua_code != NULL)
+		return lua_code;
+	size_t n = strlen(modname);
+	assert(n < MAX_MODNAME);
+	char fullname[MAX_MODNAME];
+	strlcpy(fullname, modname, n + 1);
+
+	/* 2. slow path: process as `@builtin/%s.lua` */
+	/* check prefix first */
+	const char *prefix = "@builtin/";
+	const size_t prefix_len = strlen(prefix);
```

Comment:
Good question, and I have a story... 
Here is the dilemma I've faced - certainly, as an old-boy the 1st intention was to write a couple of macro definitions with appropriate `sizeof(x) - 1` in some context to make sure that this would be real constant expression at the run-time. But `checkpatch` doesn't like macros very much, so, assuming that any modern compiler in release mode should easily distinguish constant expression in such context I've simplified this to the current version.

Even without `strlen` being treated as `constexpr` (as it would be treated in similar cases in C++) it's easily calculated at compile time in C and corresponding code using literal constants generated, even for quite old compilers. Please see code for my gcc 8.3 from ancient Debian - https://godbolt.org/z/zq34oGGvv

![image](https://user-images.githubusercontent.com/4956130/199229428-3fd31c3c-0dc4-4c35-a5fc-2cd7b59e20dc.png)

So even the dumbest, simplest code, written mainly for clarity could produce optimal code in a release mode. So I'd not worry about that seeming lack of efficiency, it's quite opposite.